### PR TITLE
MWPW-158345 Decode CaaS Config

### PR DESCRIPTION
**Pull request by Brandon32**
* Decode CaaS Config from query index, #246 
  * Press Fetch index to run report

Resolves: [MWPW-158345](https://jira.corp.adobe.com/browse/MWPW-158345)

**Test URLs:**
- Before: https://main--bacom--adobecom.hlx.live/drafts/bmarshal/url-decode?martech=off
- After: https://bmarshal-valid-caas--bacom--adobecom.hlx.live/drafts/bmarshal/url-decode?martech=off

### CODE REVIEW

File: blocks/url-decode/url-decode.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+const { createTag, getConfig, parseEncodedConfig } = await import(`${LIBS}/utils/utils.js`);
+const { decodeCompressedString } = await import(`${LIBS}/blocks/caas/utils.js`);
+
+const URL_COLUMN = 'caas-url';
+
+/* c8 ignore next */
+const delay = (milliseconds) => new Promise((resolve) => { setTimeout(resolve, milliseconds); });
+
+export const loadQueryIndex = async (url) => {
```
Comment (JasonHowellSlavin) [inline blocks/url-decode/url-decode.js:11]: Should see if anything like this exists in Milo as a util somewhere, if not you should definitely add it. Great way to batch and get the query index data.

Comment (Brandon32) [reply blocks/url-decode/url-decode.js:11]: Article feed is the closest we have in Milo, I took this from [bulk-updater](https://github.com/adobecom/bulk-updater/blob/main/bulk-update/bulk-update.js#L21-L45).

File: blocks/url-decode/url-decode.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  try {
+    data = await loadQueryIndex(queryIndex.href);
+  } catch (e) {
+    /* c8 ignore next 4 */
+    window.lana?.log(`Error fetching data from url: ${queryIndex.href}`, { tags: 'info,url-decode' });
+    report.append(createTag('p', { class: 'error' }, 'Error fetching data'));
+    return;
+  }
+
+  const decodedReports = data.map(async (page) => {
```
[the remark this answers is not in the document]
Comment (Brandon32) [reply blocks/url-decode/url-decode.js]: Had to re-write this and pulled it out to it's own function.

File: blocks/url-decode/url-decode.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+const sortValidLinks = (a, b) => {
+  if (a.validLink === b.validLink) {
+    return 0;
+  } if (a.validLink === 'Yes') {
+    return -1;
+  }
+  return 1;
+};
+
+async function fetchConfig(el, configColumn) {
```
[the remark this answers is not in the document]
Comment (Brandon32) [reply blocks/url-decode/url-decode.js]: Agreed, generateReport sounds good.

File: test/blocks/url-decode/url-decode.test.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    const options = select.querySelectorAll('option');
+    expect(select).to.exist;
+    expect(options).to.have.length.greaterThan(2);
+  });
+
+  it('shows "Fetch Index" button', async () => {
+    const el = document.querySelector('.url-decode');
+    const button = el.querySelector('button[type="submit"]');
+    expect(button).to.exist;
+  });
```
[the remark this answers is not in the document]
Comment (Brandon32) [reply test/blocks/url-decode/url-decode.test.js:44]: Sounds good, thanks

**Review by meganthecoder — APPROVED** [review]

**Pull request merged**